ddl/ddl_api.go

```diff
@@ -2343,7 +2343,9 @@ func modifiableCharsetAndCollation(toCharset, toCollate, origCharset, origCollat
 	if !charset.ValidCharsetAndCollation(toCharset, toCollate) {
 		return ErrUnknownCharacterSet.GenWithStack("Unknown character set: '%s', collation: '%s'", toCharset, toCollate)
 	}
-	if toCharset == charset.CharsetUTF8MB4 && origCharset == charset.CharsetUTF8 {
+	if (origCharset == charset.CharsetUTF8 && toCharset == charset.CharsetUTF8MB4) ||
+		(origCharset == charset.CharsetUTF8 && toCharset == charset.CharsetUTF8) ||
+		(origCharset == charset.CharsetUTF8MB4 && toCharset == charset.CharsetUTF8MB4) {
```

Comment:
this change is needless, because L2353 has check the case that toCharset is same with origCharset.
https://github.com/pingcap/tidb/pull/10958/files#diff-edfa54cbe2fdb418f0a77bd0ec4a56f2R2353

Reply:
hi @Deardrops L2353 is used to report the error message. If the check is passed(2346 ~ 2348), `nil` will be returned, and the change is used to allow changing collate when the charset is not changed.

Reply:
Sorry I point out a wrong place before. Not in L2353, but in [L2361](https://github.com/pingcap/tidb/pull/10958/files/e23ba9e480618ed994f176cd4f7fdbb2f02b850d#diff-edfa54cbe2fdb418f0a77bd0ec4a56f2R2361), there is also a `return nil` for the case that toCharset is same with origCharset.

Reply:
Sorry I point out a wrong place before. Not in L2353, but in [L2361](https://github.com/pingcap/tidb/pull/10958/files/e23ba9e480618ed994f176cd4f7fdbb2f02b850d#diff-edfa54cbe2fdb418f0a77bd0ec4a56f2R2361), there is also a `return nil` for the case that toCharset is same with origCharset.

Reply:
The code is that we allow changing the collation while keeping charset unchanged if the charset is utf8/utf8mb4.

Please check the test case https://github.com/pingcap/tidb/pull/10958/files/e23ba9e480618ed994f176cd4f7fdbb2f02b850d#diff-703ae6b7872b425273d1832c198598c8R1751 for this logic. @Deardrops